### PR TITLE
RM#70266: create batch to manage stockRules

### PR DESCRIPTION
**Pull request by oma-axelor**

### CODE REVIEW

File: axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            sll.add(stockLocationLine);
+            stockLocationLineByStockRules.put(stockRules, sll);
+          }
+        }
+      }
+    }
+    return stockLocationLineByStockRules;
+  }
+
+  private List<StockRules> getToBeCheckedStockRules(StockLocation stockLocation) {
```
Comment (kra-axelor) [inline axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java:145]: Set visibility to proteceted please

File: axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    SupplychainBatch supplychainBatch = batch.getSupplychainBatch();
+    Set<StockRules> usedStockRules = supplychainBatch.getUsedStockRulesSet();
+    return ObjectUtils.notEmpty(usedStockRules)
+        ? usedStockRules.stream()
+            .filter(sr -> stockLocationAndItsParents.contains(sr.getStockLocation()))
+            .collect(Collectors.toList())
+        : getStockRulesOfStockLocationsAndUseCaseStockControl(stockLocationAndItsParents);
+  }
+
+  private List<StockRules> getStockRulesOfStockLocationsAndUseCaseStockControl(
```
Comment (kra-axelor) [inline axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java:157]: same as above

File: axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        for (StockLocationLine stockLocationLine : stockLocationLines) {
+          try {
+            processStockLocationLineNonCompliantToStockRules(stockRule, stockLocationLine);
+            incrementDone();
+          } catch (Exception e) {
+            incrementAnomaly();
+            TraceBackService.trace(e, null, batch.getId());
+          }
+        }
+        offset += AbstractBatch.FETCH_LIMIT;
```
Comment (kra-axelor) [inline axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java:90]: Please manage cache here (adding a Jpa.clear() should be enough I think ?), need to test again then with more than FETCH_LIMIT entries

**Review by kra-axelor — CHANGES_REQUESTED** [review]

File: axelor-stock/src/main/java/com/axelor/apps/stock/service/StockRulesServiceImpl.java [hunk 4]
```diff
@@ -159,31 +127,87 @@ public boolean useMinStockRules(
   @Override
   public StockRules getStockRules(
       Product product, StockLocation stockLocation, int type, int useCase) {
+    Set<Long> stockLocationIds =
+        Beans.get(StockLocationService.class)
```
[the remark this answers is not in the document]
Comment (oma-axelor) [reply axelor-stock/src/main/java/com/axelor/apps/stock/service/StockRulesServiceImpl.java:131]: I do believe changing this one will result in a circular dependency problem, i will check that.

[the remark this answers is not in the document]
Comment (oma-axelor) [reply axelor-stock/src/main/java/com/axelor/apps/stock/service/StockRulesServiceImpl.java:131]: Okey it actually works.

**Pull request merged**